metrics_sdk/lib/opentelemetry/sdk/metrics/state/metric_store.rb

```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module SDK
+    module Metrics
+      module State
+        class MetricStore
+          def initialize
+            @mutex = Mutex.new
+            @epoch_start_time = Time.now.to_i
+            @epoch_end_time = nil;
+            @metric_streams = []
+          end
+
+          def collect
+            # this probably needs to take the mutex, take a snapshot of state, (reset state?), release the mutex
+            @mutex.synchronize do
+              [
+                @end_time = Time.now.to_i,
+                @metric_streams,
+                @epoch_start_time = @end_time,
+              ]
+              @metric_streams
+            end
+          end
+
+          def add_metric_stream(metric_stream)
```

Comment:
You might want to take the mutex here too?

Reply:
Yep - you could end up having this called in multiple threads simultaneously and end up dropping some metric streams.

Reply:
Updated!